server.c

```diff
@@ -0,0 +1,211 @@
+
+	//po otwarciu połączenia wysyłamy do klienta powitanie
+	sendResponse(client.socket, "220 (Welcome to STM32F4 FTP Server!) \r\n");
+
+	char *request_token;
+	//w petli odbieramy requesty od klienta
+	while (1)
+	{
+		recv(client.socket, client_request, BUFSIZ, 0);
+		// analizujemy request token po tokenie
+		// odcinamy pierwszą część (RETR, PASV, USER itp.)
+		request_token = strtok(client_request, " \r\n");
+		printf("%s\n", client_request);
+		// obsluga komendy USER
+		if (strcmp(request_token, "USER") == 0)
+		{
+			request_token = strtok(NULL, " \r\n");
+			if (request_token != NULL)
+			{
+				printf("%s\n", request_token);
+				strcpy(client.name, request_token);
+				sprintf(client_response, "230 (Hi %s!) \r\n", request_token);
+				sendResponse(client.socket, client_response);
+			}
+		}
+		// obsluga komendy PASV
+		if (strcmp(request_token, "PASV") == 0)
+		{
+			//trzeba otworzyc socket na porcie randomowym od 10k do 11k
+			data_socket = socket(AF_INET, SOCK_STREAM, 0);
+			data_addr.sin_family = AF_INET;
+			inet_pton(AF_INET, "127.0.0.1", &(data_addr.sin_addr));
+			//bierzemy DATA_PORT a potem inkrementujemy o jeden
+			do
+			{
+				data_addr.sin_port = htons(DATA_PORT);
+				DATA_PORT += 1;
+			} while (bind(data_socket, (struct sockaddr *)&data_addr, sizeof(data_addr)) < 0);
+
+			//wyciagamy z data_addr poszczegolne bajty
+			uint32_t ip_addr = ntohl(data_addr.sin_addr.s_addr);
+			uint16_t port = ntohs(data_addr.sin_port);
+			uint8_t b1 = (ip_addr >> 24), b2 = (ip_addr >> 16) % 256, b3 = (ip_addr >> 8) % 256, b4 = ip_addr % 256, b5 = port >> 8, b6 = port % 256;
+			sprintf(client_response, "227 Entering Passive Mode (%d,%d,%d,%d,%d,%d).\r\n", b1, b2, b3, b4, b5, b6);
+			sendResponse(client.socket, client_response);
+
+			//czekamy na połaczenie na portcie danych
+			listen(data_socket, 3);
+			printf("Waiting for data connection \n");
+			client.data_socket = accept(data_socket, (struct sockaddr *)&data_client_addr, (socklen_t *)&c);
+			printf("Data connection established\n");
+		}
+
+		// obsluga komendy QUIT
+		if (strcmp(request_token, "QUIT") == 0)
+		{
+			sprintf(client_response, "221 (Goodbye %s!) \r\n", client.name);
+			sendResponse(client.socket, client_response);
+			close(client.socket);
+
+			int *res = malloc(1 * sizeof(int));
+			*res = 0;
+			return (void *)res;
+		}
+
+		//obsluga komendy LIST
+		if (strcmp(request_token, "LIST") == 0)
+		{
+			sprintf(client_response, "150 Opening ASCII mode data connection for /bin/ls. \r\n"); // czy to jest dobra odpowiedz ??
+			sendResponse(client.socket, client_response);
+			request_token = strtok(NULL, " \r\n");
+			if (request_token == NULL)
+			{
+			   char cwd[256];
+   			   getcwd(cwd, sizeof(cwd));
+			   request_token = &cwd[0];
+			}
+			DIR *dir = opendir(request_token);
+			if (dir == NULL) 
+			{
+				printf("Wrong directory\n"); // tu chyba cos innego ma odpowiadac ale nie wiem co
+				sprintf(client_response, "Wrong directory \r\n");
```

Comment:
Client FTP odczytuje tylko ten trzycyfrowy kod, więc odpowiedź bez tego kodu chyba może spowodować nawet jakiś błąd.
Zdaje się że jak nie ma takiej dir to serwer po prostu kończy wypisywanie i zamyka połączenie od danych.
![image](https://user-images.githubusercontent.com/27630111/98105193-e1ac6a00-1e97-11eb-9069-561ff316d620.png)